vnext/Microsoft.ReactNative/IReactModuleBuilder.cpp

```diff
@@ -24,6 +24,54 @@ void ReactModuleBuilder::AddConstantProvider(ConstantProviderDelegate const &con
   m_constantProviders.push_back(constantProvider);
 }
 
+#ifdef DEBUG
+// Starting with RNW 0.64, native modules are called on the JS thread.
+// Modules will usually call Windows APIs, and those APIs might expect to be called in the UI thread.
+// Developers can dispatch work to the UI thread via reactContext.UIDispatcher().Post(). When they fail to do so,
+// cppwinrt will grab the failure HRESULT (RPC_E_WRONG_THREAD), convert it to a C++ exception, and throw it.
+// However, native module methods are noexcept, meaning the CRT will call std::terminate and not propagate exceptions up
+// the stack. Developers are then left with a crashing app and no good way to debug it. To improve developers'
+// experience, we can replace the terminate handler temporarily while we call out to the native method. In the terminate
+// handler, we can inspect the exception that was thrown and give an error message before going down.
+struct Terminator final {
+  Terminator() {
+    m_oldHandler = std::get_terminate();
+    std::set_terminate([]() {
+      auto ex = std::current_exception();
+      if (ex) {
+        try {
+          std::rethrow_exception(ex);
+        } catch (const winrt::hresult_error &hr) {
+          wchar_t buf[1024] = {};
+          StringCchPrintf(
+              buf,
+              ARRAYSIZE(buf),
+              L"An unhandled exception (0x%x) occurred in a native module. The exception message was:\n\n%s",
+              hr.code(),
+              hr.message().c_str());
+          auto messageBox = reinterpret_cast<decltype(&MessageBoxW)>(
+              GetProcAddress(GetModuleHandle(L"ext-ms-win-ntuser-dialogbox-l1-1-0.dll"), "MessageBoxW"));
```

Comment:
>GetModuleHandle [](http://example.com/codeflow?start=29&length=15)

Should we handle cases when we cannot load the DLL?

Reply:
The DLL should always be loaded already (via apiset, at least on onecoreuap). There isn't much we can do other than continue to crash, do you have any ideas of what else we could do?